### PR TITLE
Alias support for Key Policies

### PR DESCRIPTION
**Pull request by harshit777**
Added alias support for KeyPolicies management

### CODE REVIEW

File: ibm/service/kms/resource_ibm_kms_key_alias_test.go [hunk 1]
```diff
@@ -111,6 +111,15 @@ func TestAccIBMKMSResource_Key_Alias_Key_Check(t *testing.T) {
 					resource.TestCheckResourceAttr("ibm_kms_key_alias.testAlias", "alias", aliasName2),
 				),
 			},
+			{
+				Config: testAccCheckIBMKmsResourceAliasWithExistingAlias(instanceName, keyName, aliasName, aliasName2),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("ibm_kms_key.test", "key_name", keyName),
+					resource.TestCheckResourceAttr("ibm_kms_key_alias.testAlias", "alias", aliasName),
+					resource.TestCheckResourceAttr("ibm_kms_key_alias.testAlias2", "existing_alias", aliasName),
+					resource.TestCheckResourceAttr("ibm_kms_key_alias.testAlias2", "alias", aliasName2),
```
Comment (dinesh-venkatraman-g) [inline ibm/service/kms/resource_ibm_kms_key_alias_test.go:120]: the attribute name `ibm_kms_key_alias.testAlias2` is repeating

File: website/docs/d/kms_key_policies.html.markdown [hunk 2]
```diff
@@ -26,13 +31,15 @@ The following arguments are supported:
 
 - `endpoint_type` - (Optional, String) The type of the public or private endpoint to be used for fetching keys.
 - `instance_id` - (Required, string) The keyprotect instance guid.
-- `key_id` - (Required, string) The id of the key.
+- `key_id` - (Required - If the alias is not provided, String) The id of the key.
```
Comment (dinesh-venkatraman-g) [inline website/docs/d/kms_key_policies.html.markdown:34]:
```suggestion
- `key_id` - (Required - if the alias is not provided, String) The id of the key.
```

File: website/docs/d/kms_key_policies.html.markdown [hunk 3]
```diff
@@ -26,13 +31,15 @@ The following arguments are supported:
...
+- `alias`  - (Required - If the key_id is not provided, String) The alias of the key.
```
Comment (dinesh-venkatraman-g) [inline website/docs/d/kms_key_policies.html.markdown:35]:
```suggestion
- `alias`  - (Required - if the key_id is not provided, String) The alias of the key.
```

File: website/docs/r/kms_key_alias.html.markdown [hunk 4]
```diff
@@ -39,17 +45,18 @@ resource "ibm_cos_bucket" "smart-us-south" {
...
 
 ## Argument reference
-Review the argument references that you can specify for your resource. 
+Review the argument references that you can specify for your resource.
 
 - `alias` - (Required, Forces new resource, String) The alias name of the key.
 - `endpoint_type` - (Optional, Forces new resource, String) The type of the public endpoint, or private endpoint to be used for creating keys.
 - `instance_id` - (Required, Forces new resource, String) The hs-crypto or key protect instance GUID.
-- `key_id` - (Required, String) The key ID for which alias has to be created.
+- `existing_alias` - (Required, If the key_id is not provided, String) Existing Alias of the key.
```
Comment (dinesh-venkatraman-g) [inline website/docs/r/kms_key_alias.html.markdown:58]:
```suggestion
- `existing_alias` - (Required - if the key_id is not provided, String) Existing Alias of the key.
```

File: website/docs/r/kms_key_alias.html.markdown [hunk 5]
```diff
@@ -39,17 +45,18 @@ resource "ibm_cos_bucket" "smart-us-south" {
...
+- `key_id` - (Required, If the alias is not provided, String) The key ID for which alias has to be created.
```
Comment (dinesh-venkatraman-g) [inline website/docs/r/kms_key_alias.html.markdown:59]:
```suggestion
- `key_id` - (Required - if the alias is not provided, String) The key ID for which alias has to be created.
```

File: website/docs/r/kms_key_policies.html.markdown [hunk 6]
```diff
@@ -40,13 +46,26 @@ resource "ibm_kms_key_policies" "key_policy" {
...
+    }
+}
 ```
 
 ## Argument reference
 
 The following arguments are supported:
 
 - `endpoint_type` - (Optional, String) The type of the public or private endpoint to be used for fetching policies.
+- `key_id` - (Required - If the alias is not provided, String) The ID of the key.
```
Comment (dinesh-venkatraman-g) [inline website/docs/r/kms_key_policies.html.markdown:67]:
```suggestion
- `key_id` - (Required - if the alias is not provided, String) The ID of the key.
```

File: website/docs/r/kms_key_policies.html.markdown [hunk 7]
```diff
@@ -40,13 +46,26 @@ resource "ibm_kms_key_policies" "key_policy" {
...
+- `alias` - (Required - If the key_id is not provided, String) The alias created for the key.
```
Comment (dinesh-venkatraman-g) [inline website/docs/r/kms_key_policies.html.markdown:68]:
```suggestion
- `alias` - (Required - if the key_id is not provided, String) The alias created for the key.
```

**Review by dinesh-venkatraman-g — CHANGES_REQUESTED** [review]
Attach the test results.

File: ibm/service/kms/data_source_ibm_kms_key_policies.go [hunk 8]
```diff
@@ -160,6 +167,7 @@ func dataSourceIBMKMSKeyPoliciesRead(context context.Context, d *schema.Resource
 		instanceID = CrnInstanceID[len(CrnInstanceID)-3]
 	}
 	endpointType := d.Get("endpoint_type").(string)
+	alias := d.Get("alias").(string)
```
[the remark this answers is not in the document]
Comment (harshit777) [reply ibm/service/kms/data_source_ibm_kms_key_policies.go:170]: Nope, both the params are not allowed (Exactly one of) require either of the input from the user and will output error if both are used. for ex.
- Invalid combination of arguments: "key_id": only one of `alias,key_id` can be specified, but `alias,key_id` were specified.
- Invalid combination of arguments: "alias": only one of `alias,key_id` can be specified, but `alias,key_id` were specified.
2022/05/19 11:20:40 [ERROR] <root>: eval: *terraform.EvalSequence, err: 2 problems:

This comes as an output error when used in combination of both the inputs

resource "ibm_kms_key_policies" "root" {
  instance_id = ibm_resource_instance.kms_instance.id
  key_id = ibm_kms_key.root.key_id
  alias = ibm_kms_key_alias.alias.alias
  rotation {
    interval_month = 3
  }
  dual_auth_delete {
    enabled = false
  }
}

File: ibm/service/kms/resource_ibm_kms_key_alias_test.go (hunk 1 above)
Comment (harshit777) [reply ibm/service/kms/resource_ibm_kms_key_alias_test.go:120]: The two attributes are testing two different variables `existing_alias` and `alias`

**Comment by harshit777** [comment]
Screenshot 2022-05-20 at 4 49 37 PM

Screenshot 2022-05-20 at 4 50 44 PM

Screenshot 2022-05-20 at 4 58 22 PM

Screenshot 2022-05-20 at 4 59 15 PM

File: ibm/service/kms/data_source_ibm_kms_key_policies.go (hunk 8 above)
Comment (hkantare) [inline ibm/service/kms/data_source_ibm_kms_key_policies.go:170]: Can we use d.GetOK for both key_id and alias
```
var id string
	if v, ok := d.GetOk("key_id");ok {
		id =v.(string)
	} 
	if v, ok := d.GetOk("alias");ok {
		id = v.(string)
		key, err := api.GetKey(context, id)
		if err != nil {
			return diag.Errorf("Failed to get Key: %s", err)
		}
		d.Set("key_id", key.ID)
	}
	```

File: ibm/service/kms/resource_ibm_kms_key_alias.go [hunk 9]
```diff
@@ -86,7 +94,14 @@ func resourceIBMKmsKeyAliasCreate(d *schema.ResourceData, meta interface{}) erro
 
 	aliasName := d.Get("alias").(string)
 	keyID := d.Get("key_id").(string)
-	stkey, err := kpAPI.CreateKeyAlias(context.Background(), aliasName, keyID)
+	aliasID := d.Get("existing_alias").(string)
+	var id string
```
Comment (hkantare) [inline ibm/service/kms/resource_ibm_kms_key_alias.go:98]: use the same concept of d.GetOk for optional fields in all files

**Comment by hkantare** [comment]
And share the testcases results of the reosurces and datasources

**Pull request merged**